enhancements/insights/insights-runtime-extractor.md

```diff
@@ -0,0 +1,543 @@
+
+From the operator perspective, the insights-runtime-extractor is a black box, and its requirements are:
+
+* Determine the Pod name of each insights-runtime-extractor’s Pod (as the insights-runtime-extractor is deployed as a DaemonSet, there is a single pod instance on each worker node) and map it to its worker node.
+  * There is no service exposing the insights-runtime-extractor and its pods are accessed directly throug their DNS records.
+* When the operator starts gathering workload information, query the `HTTP GET /gather-runtime-info` request on each insights-runtime-extractor pod. The HTTP response will return a JSON payload in the standard output that represents a tree of `namespace/pod-name/container-id/runtime-info`.
+* As the operator iterates over the containers in the cluster, check if there is a runtime-info entry for the `namespace/pod-name/container-id` and add it to the existing `workloadContainerShape` (after hashing all the values)
+
+The insights-runtime-extractor functionality is provided by an image that must be pullable from the `openshift-insights` namespace and present in the cluster at installation.
+
+#### Variation and form factor considerations [optional]
+
+This enhancement to the Insight Operator is not impacting the Control Plane of OpenShift and should be available regardless of the  form of OpenShift (Standalone OCP,  Microshift, Hypershift).
+
+This enhancement only gather runtime info from Linux worker nodes and will not touch Windows worker nodes. The DaemonSet will be configured to only be deployed on Linux worker nodes (having the label `beta.kubernetes.io/os=linux`).
+
+### API Extensions
+
+This proposal does not require any API Extensions to the OpenShift Insights Operator.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+#### Hypershift [optional]
+
+This proposal only aims to gather data about the Data plane and will not inspect any containers running in the Control Plane.
+
+#### Standalone Clusters
+
+This proposal is relevant for standalone clusters.
+
+#### Single-node Deployments or MicroShift
+
+This proposal affects single-node deployment as there is additional resource consumption
+when the Insights Operator gathers workload data that will affect the only worker node.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Container Scanner API
+
+The Insights Operator communicates with the insights-runtime-extractor pods deployed by its DaemonSet with a simple HTTP server that has the following API:
+
+* `GET /gather-runtime-info`
+  * Inspect the all the containers running on the worker node. The HTTP response returns a JSON payload with the extracted data organized in a tree of `namespace/pod-name/container-id/runtime-info`:
+
+```json
+{
+  "my-namespace": {
+    "my-pod-1": {
```

Comment:
When I look at your integration work now, I wonder if we need this structure. Are the namespace and Pod names needed? 

Another question is whether we could use `imageID` as a key for the below structure? I guess it could help avoid some duplication (especially on the "aggregation" level in the Insights Operator). - e.g when I deploy my app with 3 replicas then I will have three various containerIDs (on various nodes probably) comparing to the same `imageID`

Reply:
I don't think `imageID` is a good key to identify workloads. Eg you could have the same `fedora` image that runs very different workloads based on the `args` (or environment variables)